### ISSUE TITLE
Null propagation expression does not allow `!` as a nested operator form

### ISSUE THREAD

**Issue by agocke** [dotnet/csharplang#3393]
See https://github.com/dotnet/docs/issues/17988

This means that the only correct parse of

```C#
var x = a?.b.c!.d.e;
```

is `(a?.b.c)!.d.e`, which prevents usage of null-forgiving in a chain of null-propagating accesses. To remedy this we could accept `!` as an optional token after every simple name in the null-propagating non-terminal.

cc @jcouv @gafter

**Comment by AntonLapounov** [comment]
I was not aware of this issue and filed dotnet/roslyn#43659.

**Comment by RikkiGibson** [comment]
It doesn't feel like accepting an optional token after a simple name is sufficient for a case like `Prop?.Method1()!.Method2()`. Could we solve this by considering `!.` a distinct operator which accesses a member and suppresses the nullable warning on the receiver?

**Comment by jaredpar** [comment]
Looks like the [NRT Spec](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/nullable-reference-types-specification.md) needs to be updated. Presently it says

> The postfix ! operator has no runtime effect - it evaluates to the result of the underlying expression. Its only role is to change the null state of the expression, and to limit warnings given on its use.

That is not true though cause it changes the manner in which conditional expressions are parsed hence their runtime behavior. Here is a simple [sharplab sample](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATARgFgAoEvAZgAIBvEy+y7fANkYBZKAFCGAC3wAUAYXyUADgEoadBrIBuEBJQB2EMHEoBecQH4AdENx6AcmrgBuGbPpMAnANXqJl4rIC+Vyp6atsHbny4wqKS0q7W9ApKjhraYvqGAIQmZi4RNvj2Mc6eHsR5ZLiUIiS04TZUhsVF1JQA5nAw5pRQjc15BcR41aXeVEwADJSm6jT1bS1tHUA) that demonstrates this.

**Comment by 333fred** [comment]
Well, that's how it _should_ be. This is a bug that we need to fix.

**Comment by gafter** [comment]
> That is not true though cause it changes the manner in which conditional expressions are parsed hence their runtime behavior.

Yes.  Also parentheses have runtime behavior for the same reason.

**Comment by RikkiGibson** [comment]
If nobody has written a syntax proposal for this I would be happy to take a stab at it.

**Comment by 333fred** [comment]
@RikkiGibson iirc https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#null-conditional-operator is the place we need to adjust: specifically, `null_conditional_operations` need to be expanded.

**Comment by MadsTorgersen** [comment]
The initial example was not supposed to parse; it's a bug that we accept it at all:
``` c#
var x = a?.b.c!.d.e;
```
The reason is that `!` is only supposed to apply to a `primary_expression`, and `a?.b.c` isn't a `primary_expression` but a `unary_expression`.

So in a sense, the fact that we accept this in C# 8.0 is a bug. However, I do agree that we should give it sensible semantics. We could enhance the [grammar](https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#null-conditional-operator) as follows:

``` antlr
null_conditional_expression
    : primary_expression null_conditional_operations_no_suppression suppression?
    ;

null_conditional_operations_no_suppression
    : null_conditional_operations? '?' '.' identifier type_argument_list?
    | null_conditional_operations? '?' '[' argument_list ']'
    | null_conditional_operations '.' identifier type_argument_list?
    | null_conditional_operations '[' argument_list ']'
    | null_conditional_operations '(' argument_list? ')'
    ;

null_conditional_operations
    : null_conditional_operations_no_suppression suppression?
    ;

suppression
    : '!'
    ;
```

The semantics then are that the "interior" `!`s get folded into the semantic "desugaring" of the null conditional expression, and have whatever effects they have on warnings there. But a top-level trailing `!` is a different beast and gets applied to the whole null conditional expression, changing its null state.
